server/permissions/UserAuth.ts

```diff
@@ -0,0 +1,71 @@
+import { IMatch } from "../models/match";
+import { ITeam, PopulatedTeam } from "../models/team";
+import { IUser } from "../models/user";
+import { Populate } from "../types";
+import { PermissionContext } from "./contexts/context";
+import { MatchContext, MatchContextParams } from "./contexts/MatchContext";
+import { TeamContext } from "./contexts/TeamContext";
+import { TourneyContext } from "./contexts/TourneyContext";
+import { UserRole } from "./UserRole";
+
+export class UserAuth {
+  protected user: IUser;
+
+  constructor(user: IUser) {
+    if (!user) {
+      throw new Error("User is not defined");
+    }
+    this.user = user;
+  }
+
+  public getUser() {
+    return this.user;
+  }
+
+  public withContext(context: PermissionContext) {
+    return new UserAuthWithContext(this.user, context);
+  }
+
+  public forMatch(match: IMatch, params: MatchContextParams) {
+    return this.withContext(new MatchContext(match, params));
+  }
+
+  public forTeam(team: Populate<ITeam, PopulatedTeam>) {
+    return this.withContext(new TeamContext(team));
+  }
+
+  public forTourney(tourney: string) {
+    return this.withContext(new TourneyContext(tourney));
+  }
+}
+
+export class UserAuthWithContext extends UserAuth {
+  private context: PermissionContext;
+  private hasSuperRole: boolean = false;
+
+  constructor(user: IUser, context: PermissionContext) {
+    super(user);
+    this.context = context;
+    this.hasSuperRole = this.user.admin;
```

Comment:
what about the checks for Host and Developer? this differs from the clientside user auth which has:
```ts
this._hasSuperRole = this._user.admin || SUPER_ROLES.some(role => this._context.hasRole(this._user, role));
```

Reply:
Oops missed this, fixed in 24c052903c0e66972aca26cce9caba37e84c31ca